### ISSUE TITLE
cors is broken when trying access directory/login

### ISSUE THREAD

**Issue by vegarringdal** [Wakanda/wakanda-issues#21]
cors is broken when trying access directory/login

I have no issues accessing data... but cant login, my script works fine as long as I dont run a node serving the pages..

Sample code that can be used with simple node static server:
https://github.com/vegarringdal/testPromise
just edit the defaultconnectSetting in the index.js file on the one runing node to you wakanda server

Tried with this also on the one running node
http_request.withCredentials = true;
http_request.setRequestHeader('Access-Control-Allow-Credentials', 'true');

But this didnt help, got this repsponse back:
MLHttpRequest cannot load http://127.0.0.1:8081/rest/$directory/login/. Response to preflight request doesn't pass access control check: Credentials flag is 'true', but the 'Access-Control-Allow-Credentials' header is ''. It must be 'true' to allow credentials. Origin 'http://127.0.0.1:3010' is therefore not allowed access.

**Comment by hamzahik** [comment]
A better explanation would help alot.

After reading your code, the situation I think is : 
- You have a nodejs server running at a port let us say 3000
- you have a wakanda server running at port 8081
- you open http://127.0.0.1:3000 on a browser
- A request for /rest/$directory/login is sent to 127.0.0.1:8081
- the browser blocks the request for CORS reasons.

First, you need to activate CORS on your Wakanda Project ( settings file ) : add 127.0.0.1:3000 with the methods you need to the authorized list.

For now support for CORS only works when you do not add special headers to your requests, so you might need to remove that header you added : Access-Control-Allow-Credentials  
But you should keep the property withCredentials to be sure the browser sends the cookies.

Just be aware that if you are using Firefox with syncronous request the withCredentials property might be ignored.

Let me know how it works out.

**Comment by vegarringdal** [comment]
I have cors working because I have enabled that on the server, I can access the data since I have no restriction on that part..
Its only when I do try to use the directory with cors the issues start coming.

I have tried adding this to my http request, but did not help

```javascript
http_request.withCredentials = true;
http_request.setRequestHeader('Access-Control-Allow-Credentials', 'true');
```

But then I get this response:
MLHttpRequest cannot load http://127.0.0.1:8081/rest/$directory/login/. Response to preflight request doesn't pass access control check: Credentials flag is 'true', but the 'Access-Control-Allow-Credentials' header is ''. It must be 'true' to allow credentials. Origin 'http://127.0.0.1:3010' is therefore not allowed access.

From the looks of it it looks like the server should have responed with Access-Control-Allow-Credentials set to "true", but its blank, that migh be the issue
From the look of it the directory with cors looks to be bugged..



Im using chrome browser btw.
Node server is just serving static pages:

code for it:
```javascript
var express = require('express');
var app = express();
app.use(express.static('public'));

var server = app.listen(3010, function () {
    var host = server.address().address;
    var port = server.address().port;

    console.log('Example app listening at http://%s:%s', host, port);
});
```

**Comment by hamzahik** [comment]
@vegarringdal Yes I can reproduce the bug now.

<img width="1080" alt="issue-21-cors" src="https://cloud.githubusercontent.com/assets/6703912/11825968/a5629262-a382-11e5-977c-703ab8a0a1b2.PNG">

**Comment by vegarringdal** [comment]
Great, but you didnt use cors there on first calls did you?
But anyways think I also found a bug when using remove by key, only get issue when using cors, Ill try and make a simple wak solution and node server so its easy to test.

**Comment by vegarringdal** [comment]
Update:
-remove by key fails too when using cors

Sample solution to test
https://dl.dropboxusercontent.com/u/23375256/wakandaRest/aurelia%20%282%29.zip

This calls project at port 8082, get error in login and delete by key
http://127.0.0.1:8081/

This calls project at port 8082-this only calls itself, so no error, just to prove code works...
http://127.0.0.1:8082/

**Comment by vegarringdal** [comment]
Looks like release of entityset is not working on cors too.

**Comment by hamzahik** [comment]
Bug fixed. It will be released in Wakanda 1.0.1
@vegarringdal thanks for the report

**Comment by hamzahik** [comment]
@vegarringdal wakanda community 1.0.1 containing the fix is available for download on wakanda.org

**Comment by vegarringdal** [comment]
Tested with the sample solution I posted :
https://dl.dropboxusercontent.com/u/23375256/wakandaRest/aurelia%20%282%29.zip

sill dont allow login even if I add
```javascript
http_request.withCredentials = true;
http_request.setRequestHeader('Access-Control-Allow-Credentials', 'true');
```

Did you guys test this with async call in a script file and not just browser console with sync call to server?
Or do credential call to server haveto be sync calls?
if you see anything wrong in the sample solution then please tell me

**Comment by hamzahik** [comment]
@vegarringdal With some modifications on your code it now works correctly.

[aurelia.zip](https://github.com/Wakanda/wakanda-issues/files/66402/aurelia.zip)

<img width="540" alt="21-1" src="https://cloud.githubusercontent.com/assets/6703912/11893330/193fdbce-a570-11e5-918c-57d78b127339.PNG">

Don't forget what I told you last time : `For now support for CORS only works when you do not add special headers to your requests`

wafCore.js, line 308 : 

```javascript
//make option
//http_request.setRequestHeader('Content-Type', 'application/json');
//http_request.setRequestHeader('If-Modified-Since', 'Thu, 1 Jan 1970 00:00:00 GMT'); // due to IE9 caching XHR
//http_request.setRequestHeader('Cache-Control', 'no-cache'); // due to IE9 caching XHR
//http_request.setRequestHeader('Access-Control-Allow-Origin', '*'); //just if case you use node in the middle... (other project Im playing around with)
			
http_request.withCredentials = true;
```

**Comment by vegarringdal** [comment]
ahh, my bad... explains why I got it to work in console..
thanks

Please add it to docs here, so others know also
* You should add this to http://doc.wakanda.org/home2.en.html#/HTTP-REST/Authenticating-Users.201-814507.en.html

*http://doc.wakanda.org/home2.en.html#/Wakanda/0.Beta/Project-Settings-File.300-684392.en.html

**Comment by hamzahik** [comment]
@vegarringdal yes you're right. The documentation is a real issue we are working on that. 
What do you think of the new approach we are trying for the documentation : 

https://wakanda.github.io/angular-wakanda

https://wakanda.github.io/wakanda-api-reference/docs/

**Comment by vegarringdal** [comment]
Looks a lot cleaner, nice.
Code samples will be important part also :-)
Just a question, have team considered publishing some benchmarks, with the static server and just rest api (for users who need database only)
Knowing servers limits will be important, as you guys improve it, will will also show this.

**Comment by hamzahik** [comment]
We have done lots of stress testing on the server especially before this
release for two reasons :
- we've changed the JavaScript engine.
- we've added the possibility to configure the number of HTTP Threads per
project.( this is being documented )

The tests we have done for the 1.0.0 release were aimed to validate the two
points above by checking : memory consumption, leaks, threads justly shared
between cors, requests queuing is done as it should, each client receives
the response for his own request..

The solutions used on those tests did not represent real world apps, that
was not the goal. The goal was to reproduce special corner case scenarios.
So the results are not really worth sharing.

We now have almost completed a simple application using the 1.0.0 release (
web & mobile ), we will base our performance tests on it and this time we
will share the results with the community :)
